### DISCUSSION TITLE
ESP32-S3 build

### DISCUSSION THREAD

**Discussion by japuzzo** [shorepine/tulipcc#8] [General]
Having issues compiling the code on my macbook.
Is there a bin file that I could grab to work with while I work out how to compile?

**Comment by bwhitman** [comment]
Yes, I will be posting the .bin during the first release, which I expect to be sometime this coming week (Feb 20-24)
Would love to hear what your issues are compiling it, so that we can help others!

**Comment by bwhitman** [comment]
I put a test build in releases just now -- give it a go, let me know how it works for you

**Comment by japuzzo** [comment]
Cool! point me to where it is, since I can't seem to find it

**Comment by bwhitman** [comment]
https://github.com/bwhitman/tulipcc/releases/tag/v0.1a

**Comment by japuzzo** [comment]
Thank you, I will report back when I get the time to apply it.

**Comment by japuzzo** [comment]
Still can not get your version of esp-idf working. But was able to install espressif version following these directions:
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/get-started/linux-macos-setup.html

But when I try to apply these two files, commend errors out:

> $ parttool.py write_partition --partition-name=vfs --input=tulip-lfs.bin
Running /Users/apuzzoj/.espressif/python_env/idf5.1_py3.8_env/bin/python /Users/apuzzoj/esp/esp-idf/components/esptool_py/esptool/esptool.py read_flash 32768 3072 /var/folders/l_/c_2tyh_133n9rgj742dpmv9c0000gn/T/tmp6f11mbz5...
esptool.py v4.5
Found 4 serial ports
Serial port /dev/cu.wchusbserial552E0070961
Connecting....
Detecting chip type... ESP32-S3
Chip is ESP32-S3 (revision v0.1)
Features: WiFi, BLE
Crystal is 40MHz
MAC: 68:b6:b3:29:9d:38
Uploading stub...
Running stub...
Stub running...
3072 (100 %)
3072 (100 %)
Read 3072 bytes at 0x00008000 in 0.3 seconds (87.8 kbit/s)...
Hard resetting via RTS pin...
Traceback (most recent call last):
  File "/Users/apuzzoj/esp/esp-idf/components/partition_table/parttool.py", line 358, in <module>
    main()
  File "/Users/apuzzoj/esp/esp-idf/components/partition_table/parttool.py", line 351, in main
    op(**common_args)
  File "/Users/apuzzoj/esp/esp-idf/components/partition_table/parttool.py", line 176, in _write_partition
    target.write_partition(partition_id, input)
  File "/Users/apuzzoj/esp/esp-idf/components/partition_table/parttool.py", line 162, in write_partition
    self.erase_partition(partition_id)
  File "/Users/apuzzoj/esp/esp-idf/components/partition_table/parttool.py", line 154, in erase_partition
    partition = self.get_partition_info(partition_id)
  File "/Users/apuzzoj/esp/esp-idf/components/partition_table/parttool.py", line 149, in get_partition_info
    raise Exception('Partition does not exist')
Exception: Partition does not exist

So how to create the partitions? Again I can not get your tools to run:

> $ python tulip_fs_create.py
Traceback (most recent call last):
  File "tulip_fs_create.py", line 6, in <module>
    from littlefs import lfs
ModuleNotFoundError: No module named 'littlefs'

Tried to insall littlefs but hit a brick wall
> $ pip install littlefs
ERROR: Could not find a version that satisfies the requirement littlefs (from versions: none)
ERROR: No matching distribution found for littlefs

**Comment by japuzzo** [comment]
So after I flashed the ESP32-S3 with MicroPython I was able to wright the  tulip FS

> $ parttool.py write_partition --partition-name=vfs --input=tulip-lfs.bin
Running /Users/apuzzoj/.espressif/python_env/idf5.1_py3.8_env/bin/python /Users/apuzzoj/esp/esp-idf/components/esptool_py/esptool/esptool.py read_flash 32768 3072 /var/folders/l_/c_2tyh_133n9rgj742dpmv9c0000gn/T/tmp8nlmdf6k...
esptool.py v4.5
Found 4 serial ports
Serial port /dev/cu.wchusbserial552E0070961
Connecting...
Detecting chip type... ESP32-S3
Chip is ESP32-S3 (revision v0.1)
Features: WiFi, BLE
Crystal is 40MHz
MAC: 68:b6:b3:29:9d:38
Uploading stub...
Running stub...
Stub running...
3072 (100 %)
3072 (100 %)
Read 3072 bytes at 0x00008000 in 0.3 seconds (87.4 kbit/s)...
Hard resetting via RTS pin...
Running /Users/apuzzoj/.espressif/python_env/idf5.1_py3.8_env/bin/python /Users/apuzzoj/esp/esp-idf/components/esptool_py/esptool/esptool.py erase_region 2097152 6291456...
esptool.py v4.5
Found 4 serial ports
Serial port /dev/cu.wchusbserial552E0070961
Connecting....
Detecting chip type... ESP32-S3
Chip is ESP32-S3 (revision v0.1)
Features: WiFi, BLE
Crystal is 40MHz
MAC: 68:b6:b3:29:9d:38
Uploading stub...
Running stub...
Stub running...
Erasing region (may be slow depending on size)...
Erase completed successfully in 19.9 seconds.
Hard resetting via RTS pin...
Running /Users/apuzzoj/.espressif/python_env/idf5.1_py3.8_env/bin/python /Users/apuzzoj/esp/esp-idf/components/esptool_py/esptool/esptool.py write_flash 2097152 tulip-lfs.bin...
esptool.py v4.5
Found 4 serial ports
Serial port /dev/cu.wchusbserial552E0070961
Connecting....
Detecting chip type... ESP32-S3
Chip is ESP32-S3 (revision v0.1)
Features: WiFi, BLE
Crystal is 40MHz
MAC: 68:b6:b3:29:9d:38
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Flash will be erased from 0x00200000 to 0x006fffff...
Compressed 5242880 bytes to 99845...
Wrote 5242880 bytes (99845 compressed) at 0x00200000 in 35.9 seconds (effective 1168.1 kbit/s)...
Hash of data verified.
Leaving...
Hard resetting via RTS pin...
Written contents of file 'tulip-lfs.bin' at offset 0x200000

But console shows issues

> The filesystem appears to be corrupted. If you had important data there, you
may want to make a flash snapshot to try to recover it. Otherwise, perform
factory reprogramming of MicroPython firmware (completely erase flash, followed
by firmware programming).

Keeps spiting that out over and over. Next if I try and put your version of micropython 

> $ parttool.py write_partition --partition-name=factory --input=micropython.bin
Running /Users/apuzzoj/.espressif/python_env/idf5.1_py3.8_env/bin/python /Users/apuzzoj/esp/esp-idf/components/esptool_py/esptool/esptool.py read_flash 32768 3072 /var/folders/l_/c_2tyh_133n9rgj742dpmv9c0000gn/T/tmpey9_g1_7...
esptool.py v4.5
Found 4 serial ports
Serial port /dev/cu.wchusbserial552E0070961
Connecting....
Detecting chip type... ESP32-S3
Chip is ESP32-S3 (revision v0.1)
Features: WiFi, BLE
Crystal is 40MHz
MAC: 68:b6:b3:29:9d:38
Uploading stub...
Running stub...
Stub running...
3072 (100 %)
3072 (100 %)
Read 3072 bytes at 0x00008000 in 0.3 seconds (87.3 kbit/s)...
Hard resetting via RTS pin...
Running /Users/apuzzoj/.espressif/python_env/idf5.1_py3.8_env/bin/python /Users/apuzzoj/esp/esp-idf/components/esptool_py/esptool/esptool.py erase_region 65536 2031616...
esptool.py v4.5
Found 4 serial ports
Serial port /dev/cu.wchusbserial552E0070961
Connecting...
Detecting chip type... ESP32-S3
Chip is ESP32-S3 (revision v0.1)
Features: WiFi, BLE
Crystal is 40MHz
MAC: 68:b6:b3:29:9d:38
Uploading stub...
Running stub...
Stub running...
Erasing region (may be slow depending on size)...
Erase completed successfully in 6.9 seconds.
Hard resetting via RTS pin...
Traceback (most recent call last):
  File "/Users/apuzzoj/esp/esp-idf/components/partition_table/parttool.py", line 358, in <module>
    main()
  File "/Users/apuzzoj/esp/esp-idf/components/partition_table/parttool.py", line 351, in main
    op(**common_args)
  File "/Users/apuzzoj/esp/esp-idf/components/partition_table/parttool.py", line 176, in _write_partition
    target.write_partition(partition_id, input)
  File "/Users/apuzzoj/esp/esp-idf/components/partition_table/parttool.py", line 170, in write_partition
    raise Exception('Input file size exceeds partition size')
Exception: Input file size exceeds partition size

Console shows it's stuck in boot loop since boot area is not there

> ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x3 (RTC_SW_SYS_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x403bb412
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fcd0108,len:0xf24
load:0x403b6000,len:0xc2c
load:0x403ba000,len:0x2d60
entry 0x403b61f4
E (29) esp_image: image at 0x10000 has invalid magic byte (nothing flashed here?)
E (29) boot: Factory app partition is not bootable
E (31) boot: No bootable app partitions in the partition table
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021

**Comment by japuzzo** [comment]
Just to make sure it's not a bad chip, I flashed micropython back on:

> $ sudo esptool.py --chip esp32s3 --port /dev/cu.wchusbserial552E0070961 erase_flash
esptool.py v4.5
Serial port /dev/cu.wchusbserial552E0070961
Connecting....
Chip is ESP32-S3 (revision v0.1)
Features: WiFi, BLE
Crystal is 40MHz
MAC: 68:b6:b3:29:9d:38
Uploading stub...
Running stub...
Stub running...
Erasing flash (this may take a while)...
Chip erase completed successfully in 23.6s
Hard resetting via RTS pin...

> $ sudo esptool.py --chip esp32s3 --port /dev/cu.wchusbserial552E0070961 write_flash -z 0 ESP32-S3-20220618-v1.19.1.bin
Password:
esptool.py v4.5
Serial port /dev/cu.wchusbserial552E0070961
Connecting....
Chip is ESP32-S3 (revision v0.1)
Features: WiFi, BLE
Crystal is 40MHz
MAC: 68:b6:b3:29:9d:38
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Flash will be erased from 0x00000000 to 0x00157fff...
Compressed 1406576 bytes to 928185...
Wrote 1406576 bytes (928185 compressed) at 0x00000000 in 81.7 seconds (effective 137.8 kbit/s)...
Hash of data verified.
Leaving...
Hard resetting via RTS pin...

Console

> ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x1 (POWERON),boot:0x8 (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fcd0108,len:0xf24
load:0x403b6000,len:0xc2c
load:0x403ba000,len:0x2d60
entry 0x403b61f4
MicroPython v1.19.1 on 2022-06-18; ESP32S3 module with ESP32S3
Type "help()" for more information.
'>>>'

**Comment by bwhitman** [comment]
@japuzzo thanks for the detailed report. I think I understand what's happening -- if you haven't done a compile you'll also need to flash Tulip's (not micropython's) partition table.  Let me replicate your issue on a fresh ESP32-S3 and update the instructions.

**Comment by japuzzo** [comment]
Another possible clue is I'm using a ESP32-S3 WROOM-1-N8R2 which has only 8mb of flash NOT 16mb like the N16R8 version

![image](https://user-images.githubusercontent.com/1309858/221010990-ff881783-b388-4ebe-9c05-783922cfc886.png)

**Comment by bwhitman** [comment]
Ah. That model will not work. We require 8MB RAM with octal flash speed to power the display. The only two models I know to successfully work so far are (both Adafruit boards, but the bare chips also should work if building your own board)
 - https://www.adafruit.com/product/5364
 - https://www.adafruit.com/product/5336

Please check out the [build guides](https://github.com/bwhitman/tulipcc/blob/main/docs/tulip_build.md) for the specs on which chip you need. Let me know if I missed something in the descriptions there. 

Here's more on which models have Octal flash. 
<img width="825" alt="image" src="https://user-images.githubusercontent.com/76612/221028396-82b6bf1b-02aa-47d8-b168-cea4f8d9d748.png">

That said, it still should have flashed to an 8MB board (the display just wouldn't work.) I will still debug how to get the partition table flashed without compiling Tulip.

**Comment by japuzzo** [comment]
Acquired a ESP32-S3 N8R8 and also a N16R8 both report back the same errors as my other board. Where standard ESP32-S3 micropython works. Would it be possible to acquire an image (bin file) via the esptool.py for me to try.
At some point bin files will be needed, just want to test my hardware.

**Comment by bwhitman** [comment]
The bin files are all in the release. You may (?) also need to compile the partition file, which is in ports/esp32s3. I'm not near my hardware now so can't help test. 

But the best way to do this is to figure out why you can't compile Tulip. While we are in early days it'll be a lot easier to get it working that way. 

Can you show us a log of your build process that doesn't work so we can help you debug? Follow the instructions on the tulip flashing page.

**Comment by jd3096-mpy** [comment]
I build the firmware success but run failed 
see [https://github.com/bwhitman/tulipcc/issues/20](url)
So I can tell you how to complie it.
Here's every step:
git clone https://github.com/bwhitman/tulipcc.git
cd tulipcc/espidf
then ./install.sh
cd tulipcc/ports/esp32
./export.sh to load esp-idf
then make BOARD=tulip4
It works.
Is there any mistake in my step?

**Reply by bwhitman** [reply]
Use these instructions

https://github.com/bwhitman/tulipcc/blob/main/docs/tulip_flashing.md

They mostly should just work on Linux. You have a few errors in yours.

**Reply by jd3096-mpy** [reply]
> Use these instructions
> 
> https://github.com/bwhitman/tulipcc/blob/main/docs/tulip_flashing.md
> 
> They mostly should just work on Linux. You have a few errors in yours.

Thanks,I use mac to fllow your every steps the build firmware,and got the same result.
Actually, I think I use linux and mac got the same bin files.
See [https://github.com/bwhitman/tulipcc/issues/20](https://github.com/bwhitman/tulipcc/discussions/url)
Thanks

**Comment by bwhitman** [comment]
@japuzzo and @jd3096-mpy can you check out the latest .bin release ? I've verified here I can flash a board without compiling using these steps: https://github.com/bwhitman/tulipcc/discussions/22
